File: mac/scripts/make_dmg_bg.py
#!/usr/bin/env python3
"""Generate the DMG background image (600×340, Retina @2x)."""

import sys
from pathlib import Path
from PIL import Image, ImageDraw, ImageFilter, ImageFont

W, H = 1200, 680   # @2x for Retina (displayed as 600×340)


def lerp_color(c1, c2, t):
    return tuple(int(c1[i] + (c2[i] - c1[i]) * t) for i in range(len(c1)))


def main(out_path: str):
    img = Image.new("RGB", (W, H))

    # Gradient background: very dark navy → deep slate
    top    = (10, 14, 30)
    bottom = (18, 10, 38)
    for y in range(H):
        t = y / H
        color = lerp_color(top, bottom, t)
        ImageDraw.Draw(img).line([(0, y), (W - 1, y)], fill=color)

    draw = ImageDraw.Draw(img)

    # Subtle grid lines
    grid_col = (30, 35, 60, 80)
    grid_img = Image.new("RGBA", (W, H), (0, 0, 0, 0))
    gd = ImageDraw.Draw(grid_img)
    step = W // 12
    for x in range(0, W, step):
        gd.line([(x, 0), (x, H)], fill=(40, 50, 90, 40))
    for y2 in range(0, H, step):
        gd.line([(0, y2), (W, y2)], fill=(40, 50, 90, 40))
    img = Image.alpha_composite(img.convert("RGBA"), grid_img).convert("RGB")
    draw = ImageDraw.Draw(img)

    # Waveform decoration (thin lines, faint)
    import math
    wave_col = (50, 100, 200, 45)
    wave_img = Image.new("RGBA", (W, H), (0, 0, 0, 0))
    wd = ImageDraw.Draw(wave_img)
    for i in range(3):
        amp   = 18 + i * 12
        freq  = 0.008 - i * 0.001
        off_y = H * 0.72 + i * 18
        pts   = []
        for x in range(0, W + 1, 2):
            y_val = off_y + math.sin(x * freq + i * 1.2) * amp
            pts.append((x, y_val))
        for j in range(len(pts) - 1):
            wd.line([pts[j], pts[j + 1]], fill=(*lerp_color((30, 80, 200), (0, 180, 255), j / W), 40), width=1)
    blur_wave = wave_img.filter(ImageFilter.GaussianBlur(radius=2))
    img = Image.alpha_composite(img.convert("RGBA"), blur_wave).convert("RGB")
    draw = ImageDraw.Draw(img)

    # App name text
    try:
        font_title = ImageFont.truetype("/System/Library/Fonts/SFNS.ttf", 52)
        font_sub   = ImageFont.truetype("/System/Library/Fonts/SFNS.ttf", 28)
    except Exception:
        try:
            font_title = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 52)
            font_sub   = ImageFont.truetype("/System/Library/Fonts/Helvetica.ttc", 28)
        except Exception:
            font_title = ImageFont.load_default()
            font_sub   = font_title

    title     = "MOSS TTS"
    subtitle  = "Drag to Applications to install"

    # Title — bright
    tw = draw.textlength(title, font=font_title) if hasattr(draw, "textlength") else 200
    draw.text(((W - tw) / 2, 40), title, font=font_title, fill=(220, 235, 255))

    # Subtitle — muted
    sw = draw.textlength(subtitle, font=font_sub) if hasattr(draw, "textlength") else 320
    draw.text(((W - sw) / 2, 108), subtitle, font=font_sub, fill=(100, 120, 180))

    # Arrow between app and Applications
    ax, ay = W // 2, H // 2 + 20
    arrow_col = (60, 100, 200)
    # Simple arrow →
    draw.line([(ax - 60, ay), (ax + 60, ay)], fill=arrow_col, width=3)
    draw.polygon([(ax + 60, ay - 10), (ax + 80, ay), (ax + 60, ay + 10)], fill=arrow_col)

    img.save(out_path)
    print(f"DMG background saved: {out_path} ({W}×{H})")


if __name__ == "__main__":
    main(sys.argv[1] if len(sys.argv) > 1 else "/tmp/dmg_bg.png")

File: Mosslanding-Sync/Mosslanding-Sync/scripts/make_dmg_bg.swift
#!/usr/bin/env swift
// Generate the DMG background image (600×340, Retina @2x = 1200×680).
// Native CoreGraphics + Core Text port of the previous Pillow-based
// make_dmg_bg.py — no third-party deps.
//
// Usage: swift scripts/make_dmg_bg.swift <output-png-path>

import AppKit
import CoreGraphics
import CoreImage
import CoreText
import Foundation
import ImageIO
import UniformTypeIdentifiers

let W: Int = 1200
let H: Int = 680

let outPath: String = CommandLine.arguments.count > 1
    ? CommandLine.arguments[1]
    : "/tmp/dmg_bg.png"

@inline(__always) func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double { a + (b - a) * t }

func rgb(_ r: Int, _ g: Int, _ b: Int, alpha: Double = 1.0) -> CGColor {
    CGColor(red: Double(r)/255, green: Double(g)/255, blue: Double(b)/255, alpha: alpha)
}

func makeContext(width: Int, height: Int) -> CGContext {
    return CGContext(
        data: nil,
        width: width, height: height,
        bitsPerComponent: 8, bytesPerRow: 0,
        space: CGColorSpaceCreateDeviceRGB(),
        bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue
    )!
}

let ctx = makeContext(width: W, height: H)
let w = CGFloat(W), h = CGFloat(H)

// Flip to top-down to mirror the Pillow coordinate system.
ctx.translateBy(x: 0, y: h)
ctx.scaleBy(x: 1, y: -1)

// ── Background gradient ───────────────────────────────────────────────────
let bgGradient = CGGradient(
    colorsSpace: CGColorSpaceCreateDeviceRGB(),
    colors: [rgb(10, 14, 30), rgb(18, 10, 38)] as CFArray,
    locations: [0.0, 1.0]
)!
ctx.drawLinearGradient(
    bgGradient,
    start: CGPoint(x: 0, y: 0),
    end:   CGPoint(x: 0, y: h),
    options: []
)

// ── Subtle grid lines (every W/12 px) ─────────────────────────────────────
let gridColor = rgb(40, 50, 90, alpha: 0.16)
ctx.setStrokeColor(gridColor)
ctx.setLineWidth(1)
let step = CGFloat(W / 12)
var x = CGFloat(0)
while x < w {
    ctx.move(to: CGPoint(x: x, y: 0))
    ctx.addLine(to: CGPoint(x: x, y: h))
    x += step
}
var y = CGFloat(0)
while y < h {
    ctx.move(to: CGPoint(x: 0, y: y))
    ctx.addLine(to: CGPoint(x: w, y: y))
    y += step
}
ctx.strokePath()

// ── Faint waveform decoration, then Gaussian-blur the layer ───────────────
let waveCtx = makeContext(width: W, height: H)
waveCtx.translateBy(x: 0, y: h)
waveCtx.scaleBy(x: 1, y: -1)
waveCtx.setLineWidth(1)
for i in 0..<3 {
    let amp = 18.0 + Double(i) * 12.0
    let freq = 0.008 - Double(i) * 0.001
    let offY = Double(H) * 0.72 + Double(i) * 18.0
    var px = 0
    var prev = CGPoint(x: 0, y: CGFloat(offY))
    while px <= W {
        let yv = offY + sin(Double(px) * freq + Double(i) * 1.2) * amp
        let cur = CGPoint(x: CGFloat(px), y: CGFloat(yv))
        let tH = Double(px) / Double(W)
        let r = lerp(30/255, 0/255,   tH)
        let g = lerp(80/255, 180/255, tH)
        let b = lerp(200/255, 255/255, tH)
        waveCtx.setStrokeColor(CGColor(red: r, green: g, blue: b, alpha: 0.16))
        waveCtx.move(to: prev)
        waveCtx.addLine(to: cur)
        waveCtx.strokePath()
        prev = cur
        px += 2
    }
}
if let waveImage = waveCtx.makeImage() {
    let ci = CIContext()
    let blurred = CIImage(cgImage: waveImage)
        .clampedToExtent()
        .applyingFilter("CIGaussianBlur", parameters: ["inputRadius": 2.0])
        .cropped(to: CGRect(x: 0, y: 0, width: W, height: H))
    if let blurredCG = ci.createCGImage(blurred, from: blurred.extent) {
        ctx.saveGState()
        ctx.scaleBy(x: 1, y: -1)
        ctx.translateBy(x: 0, y: -h)
        ctx.draw(blurredCG, in: CGRect(x: 0, y: 0, width: w, height: h))
        ctx.restoreGState()
    }
}

// ── Title + subtitle (Core Text) ──────────────────────────────────────────
func drawCenteredText(_ string: String, font: NSFont, color: CGColor, topY: CGFloat) {
    let attrs: [NSAttributedString.Key: Any] = [.font: font, .foregroundColor: NSColor(cgColor: color)!]
    let attr = NSAttributedString(string: string, attributes: attrs)
    let line = CTLineCreateWithAttributedString(attr)
    let bounds = CTLineGetBoundsWithOptions(line, .useOpticalBounds)
    let xPos = (w - bounds.width) / 2 - bounds.origin.x

    // Core Text draws upright, so undo the flip locally.
    ctx.saveGState()
    ctx.scaleBy(x: 1, y: -1)
    ctx.translateBy(x: 0, y: -h)
    // Convert top-down y to bottom-up y for the baseline.
    let baselineY = h - topY - bounds.height - bounds.origin.y
    ctx.textPosition = CGPoint(x: xPos, y: baselineY)
    CTLineDraw(line, ctx)
    ctx.restoreGState()
}

let titleFont = NSFont(name: "SF Pro Display", size: 52)
    ?? NSFont(name: "Helvetica", size: 52)
    ?? NSFont.systemFont(ofSize: 52)
let subFont = NSFont(name: "SF Pro Display", size: 28)
    ?? NSFont(name: "Helvetica", size: 28)
    ?? NSFont.systemFont(ofSize: 28)

drawCenteredText("MOSSlanding",
                 font: titleFont,
                 color: rgb(220, 235, 255),
                 topY: 40)
drawCenteredText("Drag to Applications to install",
                 font: subFont,
                 color: rgb(100, 120, 180),
                 topY: 108)

// ── Arrow between app and Applications ────────────────────────────────────
let ax = w / 2
let ay = h / 2 + 20
let arrow = rgb(60, 100, 200)
ctx.setStrokeColor(arrow)
ctx.setLineWidth(3)
ctx.move(to: CGPoint(x: ax - 60, y: ay))
ctx.addLine(to: CGPoint(x: ax + 60, y: ay))
ctx.strokePath()
ctx.setFillColor(arrow)
ctx.beginPath()
ctx.move(to: CGPoint(x: ax + 60, y: ay - 10))
ctx.addLine(to: CGPoint(x: ax + 80, y: ay))
ctx.addLine(to: CGPoint(x: ax + 60, y: ay + 10))
ctx.closePath()
ctx.fillPath()

// ── Save ──────────────────────────────────────────────────────────────────
guard let image = ctx.makeImage() else {
    FileHandle.standardError.write(Data("failed to render image\n".utf8))
    exit(1)
}
let url = URL(fileURLWithPath: outPath)
let type: CFString = UTType.png.identifier as CFString
guard let dest = CGImageDestinationCreateWithURL(url as CFURL, type, 1, nil) else {
    FileHandle.standardError.write(Data("could not open \(outPath) for write\n".utf8))
    exit(1)
}
CGImageDestinationAddImage(dest, image, nil)
if !CGImageDestinationFinalize(dest) {
    FileHandle.standardError.write(Data("could not finalize PNG \(outPath)\n".utf8))
    exit(1)
}
print("DMG background saved: \(outPath) (\(W)×\(H))")
